### PR TITLE
feat: rest zapi diff

### PR DESCRIPTION
**Pull request by rahulguptajss**
Create 2 pollers pointing to same cluster with Rest and Zapi collector but give them different datacenter. Diff is based on datacenter. Run below hidden command.
```
bin/harvest doctor diffzapirest --zapidatacenter zapi --restdatacenter rest
```

Current report:

################## Missing Labels ##############
node_labels [max_aggr_size max_vol_num max_vol_size vendor]
qtree_labels [__name__ cluster datacenter export_policy instance job oplocks security_style status volume vserver qtree]
shelf_labels [__name__ cluster datacenter instance job model module_type op_status serial_number shelf state vendor_name]
snapmirror_labels [destination_node group_type relationship_type schedule source_node last_transfer_type]
volume_labels [is_sis_volume node]
aggr_labels [type]
disk_labels []
lun_labels []
################## Missing Metrics by Object in prometheus ##############
shelf [shelf_disk_count shelf_labels shelf_new_status]
aggr [aggr_snapshot_used_percent aggr_volume_count_flexvol aggr_inode_used_percent aggr_inode_files_total aggr_raid_plex_count aggr_inode_maxfiles_possible aggr_snapshot_size_total aggr_inode_inodefile_public_capacity aggr_inode_maxfiles_available aggr_snapshot_inode_used_percent aggr_snapshot_size_available aggr_raid_disk_count aggr_inode_inodefile_private_capacity aggr_inode_files_private_used aggr_hybrid_cache_size_total aggr_inode_files_used aggr_inode_maxfiles_used aggr_snapshot_size_used aggr_snapshot_reserve_percent aggr_space_used_percent aggr_space_reserved aggr_snapshot_maxfiles_possible]
cluster [cluster_subsystem_outstanding_alerts cluster_subsystem_suppressed_alerts cluster_subsystem_new_status]
snapmirror [snapmirror_resync_successful_count snapmirror_newest_snapshot_timestamp snapmirror_total_transfer_time_secs snapmirror_update_successful_count snapmirror_resync_failed_count snapmirror_total_transfer_bytes snapmirror_update_failed_count snapmirror_break_successful_count snapmirror_break_failed_count]
disk [disk_stats_sectors_read disk_stats_sectors_written]
snapshot [snapshot_total_schedules]
environment [environment_sensor_critical_low environment_sensor_critical_high environment_sensor_threshold_value environment_sensor_status]
volume [volume_sis_compress_saved_percent volume_inode_files_used volume_sis_dedup_saved_percent volume_sis_total_saved_percent volume_snapshot_count volume_sis_dedup_saved volume_sis_total_saved volume_sis_compress_saved volume_inode_files_total]
qtree [qtree_labels qtree_soft_file_limit qtree_file_limit qtree_disk_used qtree_disk_used_pct_disk_limit qtree_threshold qtree_id qtree_soft_disk_limit qtree_disk_limit qtree_files_used]
################## Missing Metrics from dashboard ##############
aggr [aggr_snapshot_used_percent aggr_volume_count_flexvol aggr_inode_used_percent aggr_inode_files_total aggr_inode_inodefile_public_capacity aggr_snapshot_inode_used_percent aggr_snapshot_size_available aggr_raid_disk_count aggr_inode_inodefile_private_capacity aggr_inode_files_used aggr_snapshot_size_used aggr_snapshot_reserve_percent aggr_space_used_percent aggr_snapshot_maxfiles_possible]
cluster [cluster_subsystem_outstanding_alerts cluster_subsystem_suppressed_alerts cluster_subsystem_new_status]
snapmirror [snapmirror_resync_successful_count snapmirror_update_successful_count snapmirror_resync_failed_count snapmirror_total_transfer_bytes snapmirror_update_failed_count snapmirror_break_successful_count snapmirror_break_failed_count]
shelf [shelf_disk_count shelf_labels shelf_new_status]
snapshot [snapshot_total_schedules]
volume [volume_sis_compress_saved_percent volume_sis_dedup_saved_percent volume_snapshot_count volume_sis_dedup_saved volume_sis_total_saved volume_sis_compress_saved]
##################

### CODE REVIEW

File: cmd/tools/doctor/restZapiDiff.go [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+	metricDiffMap := metricDiff(zapiDataCenterName, restDataCenterName)
+	for k, v := range metricDiffMap {
+		fmt.Println(k, " : ", strings.Join(v, `','`))
+	}
+
+	fmt.Println("################## Missing Metrics from dashboard ##############")
+	dashboardDiffMap := make(map[string][]string)
+	for key, value := range metricDiffMap {
+		for _, v := range value {
+			filepath.Walk("/Users/rahulg2/GolandProjects/harvest/grafana/dashboards/cmode", getWalkFunc(v, dashboardDiffMap, key))
```
Comment (cgrinds) [inline cmd/tools/doctor/restZapiDiff.go:35]: this can be a relative path

Comment (rahulguptajss) [reply cmd/tools/doctor/restZapiDiff.go:35]: right i'll fix it.

**Comment by rahulguptajss** [comment]
Rebased it due to lint error.

**Comment by rahulguptajss** [comment]
resolved review comments.

**Review by cgrinds — APPROVED** [review]
nice job!

**Review by Hardikl — APPROVED** [review]
Good one. This will help a lot.

File: cmd/tools/doctor/restZapiDiff.go [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+
+	fmt.Println("################## Missing Metrics by Object in prometheus ##############")
+	metricDiffMap := metricDiff(zapiDataCenterName, restDataCenterName)
+	for k, v := range metricDiffMap {
+		fmt.Println(k, v)
+	}
+
+	fmt.Println("################## Missing Metrics from dashboard ##############")
+	dashboardDiffMap := make(map[string][]string)
+	for key, value := range metricDiffMap {
```
[the remark this answers is not in the document]
Comment (rahulguptajss) [reply cmd/tools/doctor/restZapiDiff.go:33]: we want to calculate the missing metrics in rest and if they are used in dashboards.

File: cmd/main/main.go [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+)
+
+func main() {
+	//p := "../../../grafana/dashboards/cmode"
+	filepath.Walk("grafana/dashboards/cmode", func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			fmt.Println(err)
+			return err
+		}
+		fmt.Println(path, info.Size())
```
[the remark this answers is not in the document]
Comment (rahulguptajss) [reply cmd/main/main.go:16]: we don't atm. We log only object. Most mappings are one to one to dashboard.

File: conf/rest/9.13.0/disk.yaml [hunk 4]
```diff
@@ -45,4 +45,5 @@ export_options:
     - shelf
     - shelf_bay
     - serial_number
-    - failed
+    - failedb
```
[the remark this answers is not in the document]
Comment (rahulguptajss) [reply conf/rest/9.13.0/disk.yaml]: corrected

File: cmd/tools/doctor/restZapiDiff.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (Hardikl) [reply cmd/tools/doctor/restZapiDiff.go:33]: got it.

**Review by Hardikl — APPROVED** [review]

**Pull request merged**